### PR TITLE
make string/bytes distinguishable in bytecode

### PR DESCRIPTION
**Pull request by hhugo**
Follow up to https://github.com/ocaml/ocaml/pull/596.
This PR move the distinction between bytes and string down to the bytecode.
This should allow the js_of_ocaml compiler to choose different runtime representation for `bytes` and `string` (when safe-string is on).

The most important changes are:
- In bytecode, `Bytes.unsafe_to_string` & `Bytes.unsafe_of_string` are no longer no-op and now call into the C runtime. The native compilation is unchanged.
- Add a new bytecode instruction for `%string_unsafe_get`, different from  the instruction used for `%bytes_unsafe_get`

### CODE REVIEW

**Comment by hhugo** [comment]
For people reading emails only, I've amended the feature description:

> In bytecode, Bytes.unsafe_to_string & Bytes.unsafe_of_string are no longer no-op and now call into the C runtime. The native compilation is unchanged.

File: bytecomp/translcore.ml
Comment (nojb) [inline bytecomp/translcore.ml:342]: Is there a good reason to keep the old names `%caml_string_set*`? I agree that backwards compatibility is generally good, but am just wondering.

File: byterun/str.c
Comment (nojb) [inline byterun/str.c]: What is the reason for keeping these `caml_string_set*` C primitives? AFAICS they are not exposed to user code and not used by the bytecode interpreter.

Comment (xclerc) [reply byterun/str.c]: I guess it is related to your previous comment, but someone might
reference them through `external`, *e.g.* [ocp-endian](https://github.com/OCamlPro/ocplib-endian/blob/master/src/endianString.cppo.ml#L94).

Comment (nojb) [reply byterun/str.c]: But they would reference the `%caml_string_set*` name, not the C primitive, right?

Comment (xclerc) [reply byterun/str.c]: Sorry, you meant "make the old OCaml primitive name point to a new C symbol"?

I think there could still be a problem, with an `.o` file generated by a previous version
of the compiler and then linked with the new runtime; however, if magic numbers are
bumped it should not be a problem (cannot remember whether magic numbers are
automatically bumped nowadays).

Comment (nojb) [reply byterun/str.c]: Yes, bumping the magic number would do the trick; I don't think supporting "old-with-new" linking is a good idea anyway.

Comment (hhugo) [reply byterun/str.c]: I've made the change.

File: bytecomp/translcore.ml
Comment (hhugo) [reply bytecomp/translcore.ml:342]: As @xclerc pointed out, a library such as [ocp-indent](https://github.com/OCamlPro/ocplib-endian/blob/master/src/endianString.cppo.ml#L94) will benefit from this backwards compatible change.

Comment (gasche) [reply bytecomp/translcore.ml:342]: Several packages were broken by a previous version of a similar change by Hongbo that turned `%string_safe_set` into `%bytes_safe_set` -- see https://github.com/ocaml/ocaml/pull/596#issuecomment-239430327 . I would be wary of changing the external-referenceable name of a primitive.

**Comment by hhugo** [comment]
I expect the tests to pass now.
Bootstrapping was required in order to remove primitives from the runtime.

**Comment by hhugo** [comment]
```
File "/home/travis/build/ocaml/ocaml/toplevel/opttopmain.ml", line 213, characters 24-29:
Error: Unbound value unset
Hint: Did you mean set?
make: *** [toplevel/opttopmain.cmx] Error 2
```
I'm not sure what this is about..

**Comment by shindere** [comment]
hhugo (2018/02/20 12:24 -0800):
> ```
> File "/home/travis/build/ocaml/ocaml/toplevel/opttopmain.ml", line 213, characters 24-29:
> Error: Unbound value unset
> Hint: Did you mean set?
> make: *** [toplevel/opttopmain.cmx] Error 2
> ```
> I'm not sure what this is about..

My bad, sorry.

Fixed on latest trunk.

**Comment by hhugo** [comment]
`Marshal.to_string` and `Marshal.to_bytes` were calling into the same c function. It is now fixed.

**Comment by nojb** [comment]
This patch looks correct to me, and it should be backwards compatible thanks to leaving the name of old externals `caml_string_set*` untouched.

However, I think the `cmo` magic number needs to be bumped due to the bytecode change.

**Comment by xavierleroy** [comment]
Warning: this PR contains a bootstrap.  (Several, actually.)  According to house regulations it will have to be merged manually by a core developer.

**Comment by trefis** [comment]
> However, I think the `cmo` magic number needs to be bumped due to the bytecode change.

Indeed, but I think it was agreed recently that we'd change all the magic numbers between each major release anyway.
For example, we've recently merged some changes to `types.ml` without changing the magic number of `cmi`s just yet.

**Comment by damiendoligez** [comment]
@hhugo could you reorder the commits (moving up the marshall commit) so we only have to deal with two bootstraps instead of three?

Also, rebase Changes (but not the bootstrap compilers, they are a nice reminder that we need to merge manually).

**Comment by hhugo** [comment]
Not sure what you meant by
> Also, rebase Changes (but not the bootstrap compilers, they are a nice reminder that we need to merge manually).

This PR now only include 2 bootstraps

**Comment by hhugo** [comment]
I guess this PR needs a formal approval and a core-dev to manual merge & redo the bootstraps.
@nojb, would you be able to do this ?

**Comment by nojb** [comment]
@hhugo Yes, will do.

**Review by nojb — APPROVED** [review]

**Comment by nojb** [comment]
Merged as #1659, thanks!

**Comment by hhugo** [comment]
Thanks for all the boostraps

**Pull request closed without merging**
